### PR TITLE
Support the '0 dimension case' for bigarrays

### PR DESCRIPTION
**Pull request by LaurentMazare**
Handle the '0 dimension' case for bigarray in the same way other numerical libraries (e.g. numpy or tensorflow) do, which is that such a bigarray holds a scalar value.

It seems that relaxing the exceptions that currently prevent this case is enough to get this working.

Note that this PR doesn't include any support for the bigarray.{} syntax, not sure how worthy this is.
It also doesn't include a Bigarray.Array0 module but I'm happy to add one if there is any need for it.

To give some context I came along this when writing some tensorflow ocaml bindings and the current workaround is to have a wrapper around the bigarray module that handles this scalar case with a bigarray with one dimension of size one. However this isn't very neat as it requires the wrapper to store whether the embedded data is a scalar or not. It would be quite nicer to have this supported by the bigarray module directly.

### CODE REVIEW

File: otherlibs/bigarray/bigarray_stubs.c [hunk 1]
```diff
@@ -1050,7 +1050,7 @@ CAMLprim value caml_ba_slice(value vb, value vind)
 
   /* Check number of indices < number of dimensions of array */
```
Comment (alainfrisch) [inline otherlibs/bigarray/bigarray_stubs.c]: Should be `<=` now.

**Comment by alainfrisch** [comment]
The request is well founded and the implementation rather straightforward and well tested.  I'm in favor of merging if nobody is against.

**Comment by alainfrisch** [comment]
Ref : http://caml.inria.fr/mantis/view.php?id=7339

**Comment by hcarty** [comment]
Could something be added to the documentation so users realize this is available/possible?

**Comment by alainfrisch** [comment]
The ability to have dimension 0 is mentioned in bigarray.mli.

**Comment by hcarty** [comment]
Ah, I missed those lines.

I would like an Array0 module and Array1.slice to go with this. I have had projects in the past which would have benefited from an Array0.

File: otherlibs/bigarray/bigarray_stubs.c (hunk 1 above)
Comment (LaurentMazare) [reply otherlibs/bigarray/bigarray_stubs.c]: Good catch, this is fixed now.

**Comment by LaurentMazare** [comment]
I've added an Array0 module as well as Array1.slice. It makes the change larger so let me know if you think it's not worth it.

**Comment by hcarty** [comment]
I think the addition of the module, slice function and genarray conversions make this PR much more useful. Thank you for adding them!

I looked over the code and like the change.

**Comment by gasche** [comment]
I merged, thanks to everyone for the comments, and for the nice pull request.

**Comment by LaurentMazare** [comment]
That was pretty fast! Thanks all.

**Comment by gasche** [comment]
Our internal continuous integration server just let me know that clang has a seemingly-legitimate complaint about the C code:

```
bigarray_stubs.c:205:16: error: comparison of unsigned expression < 0 is always false [-Werror,-Wtautological-compare]
  if (num_dims < 0 || num_dims > CAML_BA_MAX_NUM_DIMS)
      ~~~~~~~~ ^ ~
bigarray_stubs.c:1302:16: error: comparison of unsigned expression < 0 is always false [-Werror,-Wtautological-compare]
  if (num_dims < 0 || num_dims > CAML_BA_MAX_NUM_DIMS)
      ~~~~~~~~ ^ ~
2 errors generated.
```

**Comment by LaurentMazare** [comment]
Indeed I didn't noticed that num_dims is unsigned, sorry about that. Should I create a new PR to fix this ? (removing the num_dims < 0 conditions)

**Comment by gasche** [comment]
I will do the change, run the testsuite, and commit if it works.

**Comment by gasche** [comment]
Should be fixed in 014d94159dbbf086bd63e49833e78922680e0f9a .

**Pull request merged**